core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/resourcePolicies/NaiveResourceAllocator.scala

```diff
@@ -0,0 +1,192 @@
+      return (0, 0)
+
+    val recordsByExecutionId: Map[Int, List[WorkflowRuntimeStatistics]] =
+      executionHistory
+        .filter(stats => Option(stats.getStatus).contains(1.toByte))
+        .groupBy(_.getExecutionId.intValue())
+        .view
+        .mapValues(_.toList)
+        .toMap
+
+    def trimLeadingZeroPairs(
+        records: List[WorkflowRuntimeStatistics]
+    ): List[WorkflowRuntimeStatistics] =
+      records.dropWhile(record =>
+        record.getInputTupleCnt.intValue() == 0 && record.getOutputTupleCnt.intValue() == 0
+      )
+
+    val filteredResult: List[WorkflowRuntimeStatistics] =
+      recordsByExecutionId.values.toList.flatMap(trimLeadingZeroPairs)
+
+    val workingTime = if (filteredResult.nonEmpty) {
+      filteredResult
+        .map(record =>
+          (2000000000 - record.getIdleTime.doubleValue() - record.getControlProcessingTime
+            .doubleValue()) / 1000000000 / record.getNumWorkers.intValue()
+        )
+        .sum
+    } else 2
+
+    val avgInputTupleCnt = if (filteredResult.nonEmpty) {
+      filteredResult
+        .map(record => record.getInputTupleCnt.doubleValue() / record.getNumWorkers.intValue())
+        .sum / workingTime
+    } else 0
+
+    val avgOutputTupleCnt = if (filteredResult.nonEmpty) {
+      filteredResult
+        .map(record => record.getOutputTupleCnt.doubleValue() / record.getNumWorkers.intValue())
+        .sum / workingTime
+    } else 0
+
+    (avgInputTupleCnt, avgOutputTupleCnt)
+  }
+
+  private def avgHistory(region: Region): Map[PhysicalOp, Double] = {
+    val parallelizableOperators =
+      region.getOperators.filter(op =>
+        op.parallelizable && (op.suggestedWorkerNum.isEmpty || op.suggestedWorkerNum.get == 0)
+      )
+    parallelizableOperators.map(op => op -> avgHistoryPerOperator(op.id)._1).toMap
+  }
+
+  private def allocateWorkers(
+      speeds: Map[PhysicalOp, Double],
+      totNumWorkers: Int
+  ): Map[PhysicalOp, Int] = {
+    var numWorkers = speeds.map {
+      case (opId, speed) =>
+        val workersForOp = math.max(1, (speed / speeds.values.sum * totNumWorkers).toInt)
+        opId -> workersForOp
+    }
+    val restNumWorkers = totNumWorkers - numWorkers.values.sum
+
+    for (_ <- 1 to restNumWorkers) {
+      val j = numWorkers.minBy {
+        case (opId, workerCount) =>
+          workerCount.toDouble / speeds(opId)
+      }._1
+      numWorkers = numWorkers.updated(j, numWorkers(j) + 1)
+    }
+
+    numWorkers
+  }
+  def allocate(region: Region): (Region, Double) = {
+    val histories = avgHistory(region)
+
+    var estimatedCost = 0.0
+    var operators: Set[PhysicalOp] = region.getOperators
+    if (histories.values.sum != 0) {
+      val speeds: Map[PhysicalOp, Double] = histories.map {
+        case (op, throughPut) => op -> (if (throughPut != 0) 1 / throughPut else 1)
+      }
+      val numWorkers =
+        allocateWorkers(speeds, executionClusterInfo.getAvailableNumOfWorkers(region))
```

Comment:
I do not understand this algorithm. you seem to loop through each operator and allocate workers for them. but how can you know how many more operators to allocator?